### ISSUE TITLE
Container not starting with multiple raw.idmap gids assigned

### ISSUE THREAD

**Issue by stonie08** [lxc/lxd#3659]
# Required information
 * Distribution: Debian
 * Distribution version: Stretch
 * The output of "lxc info" or if that fails:
```
architectures:
  - x86_64
  - i686
driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.9.0-3-amd64
  server: lxd
  server_version: "2.16"
  storage: zfs
  storage_version: 0.6.5.9-5
```

# Issue description
According to the mailinglist archive one can assign multiple values to the config `raw.idmap` using a command like
```echo -en "both 1003 1003\nboth 1004 1004" | lxc config set mycontainer raw.idmap -```
this also works (as in one doesn't get an error) for assigning multiple gids:
```echo -en "gid 998 999\ngid 999 110" | lxc config set mycontainer raw.idmap -```

However starting the conainter fails with:
```
error: Failed to run: /snap/lxd/3346/bin/lxd forkstart mycontainer /var/snap/lxd/common/lxd/containers /var/snap/lxd/common/lxd/logs/mycontainer/lxc.conf: 
Try `lxc info --show-log mycontainer` for more info
```

Interestingly it doesn't fail with neither of the gid mappings alone.
It also works fine when mapping multiple uids like: `"uid 1000 1000\nuid 1004 1001"`
And it works fine for `"both 1000 1000\nboth 1001 1001"`

# Steps to reproduce

1. create a new containter `mycontainer`
2. create two groups `groupadd --system lxd-group1` `groupadd --system lxd-group1`
3. figure out the gids `cat /etc/group | grep lxd`
4. map the gids to the container e.g.
4.1 `echo -en "gid 999 999\ngid 998 998" | lxc config set mycontainer raw.idmap -`
5. start the container `lxc start mycontainer`

# Information to attach

 - container log (`lxc info NAME --show-log`)
```
# lxc info --show-log mycontainer                                                                                                                                                         :(
Name: mycontainer
Remote: unix://
Architecture: x86_64
Created: 2017/08/09 20:44 UTC
Status: Stopped
Type: persistent
Profiles: default, Audio, Video

Log:

            lxc 20170814100603.795 ERROR    lxc_cgfsng - cgroups/cgfsng.c:create_path_for_hierarchy:1328 - Path "/sys/fs/cgroup/systemd//lxc/mycontainer" already existed.
            lxc 20170814100603.813 ERROR    lxc_cgfsng - cgroups/cgfsng.c:cgfsng_create:1385 - No such file or directory - Failed to create /sys/fs/cgroup/systemd//lxc/mycontainer: No such file or directory
            lxc 20170814100603.813 ERROR    lxc_cgfsng - cgroups/cgfsng.c:create_path_for_hierarchy:1328 - Path "/sys/fs/cgroup/systemd//lxc/mycontainer-1" already existed.
            lxc 20170814100603.813 ERROR    lxc_cgfsng - cgroups/cgfsng.c:cgfsng_create:1385 - No such file or directory - Failed to create /sys/fs/cgroup/systemd//lxc/mycontainer-1: No such file or directory
            lxc 20170814100603.813 ERROR    lxc_cgfsng - cgroups/cgfsng.c:create_path_for_hierarchy:1328 - Path "/sys/fs/cgroup/systemd//lxc/mycontainer-2" already existed.
            lxc 20170814100603.813 ERROR    lxc_cgfsng - cgroups/cgfsng.c:cgfsng_create:1385 - No such file or directory - Failed to create /sys/fs/cgroup/systemd//lxc/mycontainer-2: No such file or directory
            lxc 20170814100603.814 ERROR    lxc_cgfsng - cgroups/cgfsng.c:create_path_for_hierarchy:1328 - Path "/sys/fs/cgroup/systemd//lxc/mycontainer-3" already existed.
            lxc 20170814100603.814 ERROR    lxc_cgfsng - cgroups/cgfsng.c:cgfsng_create:1385 - No such file or directory - Failed to create /sys/fs/cgroup/systemd//lxc/mycontainer-3: No such file or directory
            lxc 20170814100603.838 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.839 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.839 ERROR    lxc_cgfsng - cgroups/cgfsng.c:cgfsns_chown:1507 - Error requesting cgroup chown in new namespace
            lxc 20170814100603.152 WARN     lxc_conf - conf.c:lxc_delete_network:3103 - Failed to remove "vethH9TWWB" from host: Invalid argument.
            lxc 20170814100603.152 ERROR    lxc_start - start.c:__lxc_start:1358 - Failed to spawn container "mycontainer".
            lxc 20170814100603.705 ERROR    lxc_conf - conf.c:run_buffer:416 - Script exited with status 1.
            lxc 20170814100603.705 ERROR    lxc_start - start.c:lxc_fini:546 - Failed to run lxc.hook.post-stop for container "mycontainer".
            lxc 20170814100603.705 WARN     lxc_commands - commands.c:lxc_cmd_rsp_recv:177 - Command get_cgroup failed to receive response: Connection reset by peer.
            lxc 20170814100603.705 WARN     lxc_commands - commands.c:lxc_cmd_rsp_recv:177 - Command get_cgroup failed to receive response: Connection reset by peer.
            lxc 20170814100603.705 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.705 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.705 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/systemd//lxc/mycontainer-4
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.706 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/pids//lxc/mycontainer-4
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.706 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/cpu//lxc/mycontainer-4
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.706 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.706 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/devices//lxc/mycontainer-4
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.707 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/blkio//lxc/mycontainer-4
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.707 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/perf_event//lxc/mycontainer-4
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.707 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/cpuset//lxc/mycontainer-4
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.707 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.707 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/net_cls//lxc/mycontainer-4
            lxc 20170814100603.708 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.708 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.708 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/freezer//lxc/mycontainer-4
            lxc 20170814100603.708 ERROR    lxc_conf - conf.c:write_id_mapping:3291 - Invalid argument - writing id mapping
            lxc 20170814100603.708 ERROR    lxc_conf - conf.c:userns_exec_1:4608 - Error setting up child mappings
            lxc 20170814100603.708 ERROR    lxc_cgfsng - cgroups/cgfsng.c:recursive_destroy:1288 - Error destroying /sys/fs/cgroup/memory//lxc/mycontainer-4
```

**Comment by stgraber** [comment]
Can you paste /var/snap/lxd/common/lxd/logs/mycontainer/lxc.conf right after a failed start attempt?

**Comment by stonie08** [comment]
Sure, can do

Content of  `/var/snap/lxd/common/lxd/logs/mycontainer/lxc.conf`
```
lxc.cap.drop = sys_time sys_module sys_rawio mac_admin mac_override
lxc.mount.auto = proc:rw sys:rw
lxc.autodev = 1
lxc.pts = 1024
lxc.mount.entry = /dev/fuse dev/fuse none bind,create=file,optional
lxc.mount.entry = /dev/net/tun dev/net/tun none bind,create=file,optional
lxc.mount.entry = /proc/sys/fs/binfmt_misc proc/sys/fs/binfmt_misc none rbind,create=dir,optional
lxc.mount.entry = /sys/fs/fuse/connections sys/fs/fuse/connections none rbind,create=dir,optional
lxc.mount.entry = /sys/fs/pstore sys/fs/pstore none rbind,create=dir,optional
lxc.mount.entry = /sys/kernel/debug sys/kernel/debug none rbind,create=dir,optional
lxc.mount.entry = /sys/kernel/security sys/kernel/security none rbind,create=dir,optional
lxc.mount.entry = /dev/mqueue dev/mqueue none rbind,create=dir,optional
lxc.include = /snap/lxd/3346/lxc/config//common.conf.d/
lxc.logfile = /var/snap/lxd/common/lxd/logs/mycontainer/lxc.log
lxc.loglevel = warn
lxc.arch = linux64
lxc.hook.pre-start = /snap/lxd/3346/bin/lxd callhook /var/snap/lxd/common/lxd 1 start
lxc.hook.post-stop = /snap/lxd/3346/bin/lxd callhook /var/snap/lxd/common/lxd 1 stop
lxc.tty = 0
lxc.utsname = mycontainer
lxc.mount.entry = /var/snap/lxd/common/lxd/devlxd dev/lxd none bind,create=dir 0 0
lxc.seccomp = /var/snap/lxd/common/lxd/security/seccomp/mycontainer
lxc.id_map = u 0 1000000 110
lxc.id_map = g 0 1000000 110
lxc.id_map = g 110 999 1
lxc.id_map = u 111 1000111 888
lxc.id_map = g 111 1000111 888
lxc.id_map = g 999 998 1
lxc.id_map = u 1000 1001000 999999000
lxc.id_map = g 1000 1001000 999999000
lxc.network.0.type = veth
lxc.network.0.flags = up
lxc.network.0.link = lxdbr0
lxc.network.0.hwaddr = 00:16:3e:2a:d8:a9
lxc.network.0.name = eth0
lxc.rootfs.backend = dir
lxc.rootfs = /var/snap/lxd/common/lxd/containers/mycontainer/rootfs
lxc.mount.entry = /var/snap/lxd/common/lxd/shmounts/mycontainer dev/.lxd-mounts none bind,create=dir 0 0
```

**Comment by stgraber** [comment]
Okay, so I think that's a known mix of LXC bug and kernel limitation. I believe the kernel only allows up to 5 entries in the map, you have exactly 5 in the example above, but liblxc itself uses an extra one which causes the kernel limit to be hit.

We've fixed this particular problem in liblxc, allowing you one extra entry (which would be enough in your case), but that fix isn't in a stable lxc release yet.

The "edge" snap should have that fix since it's built against latest liblxc from git. But I would very strongly recommend against using the edge channel on a production system, since those are automatic builds from the current git branches that aren't receiving any kind of testing.

To check if that will fix your issue, you could do:

 - snap refresh lxd --edge
 - lxc start mycontainer

Then to go back to the safety of the stable build, do:

 - snap refresh lxd --stable

For anyone else seeing this comment, note that it's not always possible to switch back from edge to stable. But right now it is as we've not done any database schema changes since LXD 2.16 was released.

**Comment by stonie08** [comment]
Oh okay I see. Five seems to be quite an arbitrary number for this limit. 

I assume it's about these entries:
```
lxc.id_map = u 0 1000000 110
lxc.id_map = g 0 1000000 110
lxc.id_map = g 110 999 1
lxc.id_map = u 111 1000111 888
lxc.id_map = g 111 1000111 888
lxc.id_map = g 999 998 1
lxc.id_map = u 1000 1001000 999999000
lxc.id_map = g 1000 1001000 999999000
```
The limit is for each u and g entry separately?

**Comment by stgraber** [comment]
That's correct, yes. And the limit of 5 entries is indeed very arbitrary. I believe @brauner was trying to get it increased or be eliminated entirely from the kernel.

**Comment by stgraber** [comment]
Closing this issue since the problem is caused by a kernel limitation. As I mentioned above, we've now optimized liblxc to not waste an additional entry which should make it slightly better for some of our users.

There is some hope that the kernel limitation will be lifted or at least increased to a more reasonable amount, but those discussions always take a rather long time...

**Comment by brauner** [comment]
My kernel patch was merged into mainline and is to be shipped in `4.15.`:

- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=aa4bf44dc851c6bdd4f7b61b5f2c56c84dfe2ff0
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6397fac4915ab3002dc15aae751455da1a852f25
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=758f875848d78148cf9a9cdb3ff1ddf29b234056

this will bump the maximum amount of possible id mappings from `5` to `340` so this use case should be covered by this.
